Reusable git builds

philpennock opened the pull request:
This branch implements the design which I sketched out in #147 to have reusable git builds, by splitting off the `.git` dir from the build dirs, and having a log of invocations to use with a new `gimme clean`.

Fixes #147

philpennock committed:
Reusable git builds

Use a common git bare repo for fetching code, build from git in specific
directories so that each is reusable.

Cleans up some of the gunk/tech-debt I accrued earlier in version
parsing.  I _think_ this approach succeeds with everything the earlier
approach did.

This is part 1 of addressing #147 where I sketched out the design; no
log-files yet.

philpennock committed:
use logfile to track invocations, add `gimme clean`

This gives us the other part of multiple git builds, each isolated, per
issue #147.

Caveat: the use of aliases breaks the log-file's utility.  For now, I'm
handling that by not implementing aliases for git and constraining clean
to be only for git.  We can revisit if desired.

philpennock commented:
ping

philpennock commented:
ping

philpennock committed:
Merge branch 'master' into reusable_git_builds

philpennock committed:
Merge branch 'master' into reusable_git_builds

philpennock commented:
Ping.

Lyoness committed:
Merge branch 'master' into reusable_git_builds

Lyoness commented:
Thanks for the continued pings @philpennock , we're merging a few other PRs and cutting a 1.5.2 release, and then we'll circle back to this for a 1.6.0 release in the coming weeks. 

cc/ @joepvd

philpennock committed:
Merge branch 'master' into reusable_git_builds

philpennock committed:
Switch common git from --bare to --mirror

We want all references to auto-update, so --mirror is more appropriate.

This fixes `gimme tip` being very stale.

philpennock committed:
shfmt 2.2.0 vs 2.5.1 normalization

I have shfmt v2.5.1 installed; it disagrees with shfmt v2.2.0 on how to
normalize one line of this file, thus the tests passing.  I've manually
edited to what shfmt v2.2.0 apparently desires.

philpennock commented:
This is obviously going nowhere.  Closing to get it out of my list of open PRs.
